ddl/ddl_api.go

```diff
@@ -636,6 +637,108 @@ func checkConstraintNames(constraints []*ast.Constraint) error {
 	return nil
 }
 
+func checkCreatePartitionNameUnique(part *model.PartitionInfo) error {
+	if part.Type != model.PartitionTypeRange {
+		return nil
+	}
+	newPars := part.Definitions
+	set := make(map[string]struct{})
+	for _, newPar := range newPars {
+		if _, ok := set[strings.ToLower(newPar.Name)]; ok {
+			return ErrSameNamePartition.GenByArgs(newPar.Name)
+		}
+		set[strings.ToLower(newPar.Name)] = struct{}{}
+	}
+	return nil
+}
+
+// checkCreatePartitionValue checks 'less than value' must be strictly increasing for each partition.
+func checkCreatePartitionValue(part *model.PartitionInfo) error {
+	if part.Type == model.PartitionTypeRange {
+		defs := part.Definitions
+		rangeValue := defs[0].LessThan[0]
+		if strings.EqualFold(rangeValue, "MAXVALUE") && len(defs) > 1 {
+			return errors.Trace(ErrPartitionMaxvalue)
+		}
+		currentRangeValue, err := strconv.Atoi(rangeValue)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if strings.EqualFold(defs[len(defs)-1].LessThan[0], "MAXVALUE") {
+			defs = defs[:len(defs)-1]
+		}
+
+		for i := 1; i < len(defs); i++ {
+			if strings.EqualFold(defs[i].LessThan[0], "MAXVALUE") && i != len(defs)-1 {
+				return errors.Trace(ErrPartitionMaxvalue)
+			}
+			nextRangeValue, err := strconv.Atoi(defs[i].LessThan[0])
+			if err != nil {
+				return errors.Trace(err)
+			}
+			if nextRangeValue <= currentRangeValue {
+				return errors.Trace(ErrRangeNotIncreasing)
+			}
+			currentRangeValue = nextRangeValue
+		}
+	}
+	return nil
+}
+
+func buildCreateTablePartitionInfo(ctx sessionctx.Context, d *ddl, s *ast.CreateTableStmt, cols []*table.Column) (*model.PartitionInfo, error) {
+	if s.Partition == nil {
+		return nil, nil
+	}
+	pi := &model.PartitionInfo{
+		Type:   s.Partition.Tp,
+		Enable: ctx.GetSessionVars().EnableTablePartition,
+	}
+	if s.Partition.Expr != nil {
+		buf := new(bytes.Buffer)
+		s.Partition.Expr.Format(buf)
+		pi.Expr = buf.String()
+		if s.Partition.Tp == model.PartitionTypeRange {
+			for _, col := range cols {
+				name := strings.Replace(col.Name.String(), ".", "`.`", -1)
+				if _, ok := s.Partition.Expr.(*ast.ColumnNameExpr); ok {
+					// TODO check that the expression returns an integer.
+				}
+				if _, ok := s.Partition.Expr.(ast.ExprNode); ok {
+					if !(col.Tp == mysql.TypeLong || col.Tp == mysql.TypeLonglong) && fmt.Sprintf("`%s`", name) == pi.Expr {
```

Comment:
s/`!(col.Tp == mysql.TypeLong || col.Tp == mysql.TypeLonglong)`/`col.Tp != mysql.TypeLong && col.Tp != mysql.TypeLonglong`?

Reply:
@zimulala `col.Tp != mysql.TypeLong && col.Tp != mysql.TypeLonglong` the logic is wrong.

Reply:
@zimulala It's not easy to understand if you write it this way.
```
col.Tp != mysql.TypeLong || col.Tp != mysql.TypeLonglong && fmt.Sprintf("`%s`", name) == pi.Expr
```